src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,499 @@ func (s *commitLogSource) ReadData(
 		}
 	}
 
+	if iterErr := iter.Err(); iterErr != nil {
+		return nil, iterErr
+	}
+
 	for _, encoderChan := range encoderChans {
 		close(encoderChan)
 	}
 
-	// Block until all data has been read and encoded by the worker goroutines
+	// Block until all required data from the commit log has been read and
+	// encoded by the worker goroutines
 	wg.Wait()
 	s.logEncodingOutcome(workerErrs, iter)
 
-	result := s.mergeShards(int(numShards), bopts, blockSize, blopts, encoderPool, shardDataByShard)
-	// After merging shards, its safe to cache the shardData (which involves some mutation).
-	if s.shouldCacheSeriesMetadata(runOpts, ns) {
-		s.cacheShardData(ns, shardDataByShard)
+	// Merge all the different encoders from the commit log that we created with
+	// the data that is available in the snapshot files.
+	mergeStart := time.Now()
+	s.log.Infof("Starting merge...")
+	bootstrapResult, err := s.mergeAllShardsCommitLogEncodersAndSnapshots(
+		ns,
+		shardsTimeRanges,
+		snapshotFilesByShard,
+		mostRecentCompleteSnapshotByBlockShard,
+		int(numShards),
+		blockSize,
+		shardDataByShard,
+	)
+	if err != nil {
+		return nil, err
+	}
+	s.log.Infof("Done merging..., took: %s", time.Since(mergeStart).String())
+
+	return bootstrapResult, nil
+}
+
+func (s *commitLogSource) snapshotFilesByShard(
+	nsID ident.ID,
+	filePathPrefix string,
+	shardsTimeRanges result.ShardTimeRanges,
+) (map[uint32]fs.FileSetFilesSlice, error) {
+	snapshotFilesByShard := map[uint32]fs.FileSetFilesSlice{}
+	for shard := range shardsTimeRanges {
+		snapshotFiles, err := s.snapshotFilesFn(filePathPrefix, nsID, shard)
+		if err != nil {
+			return nil, err
+		}
+		snapshotFilesByShard[shard] = snapshotFiles
+	}
+
+	return snapshotFilesByShard, nil
+}
+
+func (s *commitLogSource) newShardDataByShard(
+	shardsTimeRanges result.ShardTimeRanges,
+	numShards uint32,
+) []shardData {
+	shardDataByShard := make([]shardData, numShards)
+	for shard := range shardsTimeRanges {
+		shardDataByShard[shard] = shardData{
+			series: NewMap(MapOptions{}),
+			ranges: shardsTimeRanges[shard],
+		}
+	}
+
+	return shardDataByShard
+}
+
+func (s *commitLogSource) mostRecentCompleteSnapshotByBlockShard(
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	snapshotFilesByShard map[uint32]fs.FileSetFilesSlice,
+	fsOpts fs.Options,
+) map[xtime.UnixNano]map[uint32]snapshotMetadata {
+	var (
+		minBlock, maxBlock              = shardsTimeRanges.MinMax()
+		decoder                         = msgpack.NewDecoder(nil)
+		mostRecentSnapshotsByBlockShard = map[xtime.UnixNano]map[uint32]snapshotMetadata{}
+	)
+
+	for currBlockStart := minBlock.Truncate(blockSize); currBlockStart.Before(maxBlock); currBlockStart = currBlockStart.Add(blockSize) {
+		for shard := range shardsTimeRanges {
+			// Anonymous func for easier clean up using defer.
```

Comment:
+1

Reply:
yep exactly, seemed worth it as its only a few hundred -> thousand small allocations for a process that takes minutes to complete